### PR TITLE
feat: Dataset API add `save` method

### LINKED ISSUE #166: Add a `save` method for the `APIDataSet`
## Description
We sometimes want to save some data with a REST Api (say with Django Rest Framework), most of the time issuing a POST request. So far, the `APIDataSet` is read only, with the `_save` method raising a `DataSetError`, we'd want to extend it to make a request.

## Context
I have successfully subclassed the `APIDataSet` to overwrite this method, but it could be useful to have this natively here.

## Possible Implementation
One option would be to use the same args as the `APIDataSet`, and to only define it as a saved dataset. One argument that I needed was a `chunk_size` which could be part of the `save_args` to limit the size of the payload when the data we try to save is too big and causes issues.
The data that we save is a list of dict, most of the time the result of a `pandas.DataFrame.to_dict(orient="records")`.

## Possible Alternatives
The accepted data could be a dataframe directly, but I feel like adding a node so that the users have to be explicit about their serialization would lead to less pain in the end ;).

We could hardcode the `POST` method, but I can image cases where a `PUT` method for a partial update could be useful as well.

cc @McDonnellJoseph

### PR DESCRIPTION
**Pull request by McDonnellJoseph**
## Description
We sometimes want to save some data with a REST Api (say with Django Rest Framework), most of the time issuing a POST request. So far, the APIDataSet is read only, with the _save method raising a DataSetError, we'd want to extend it to make a request.

Closes kedro-org/kedro-plugins#166 

## Development notes
If applied, this commit will:

1. Enable save method for APIDataSet, by sending packets of size chunk_size, default is 100 to server. Default HTTP method is POST and returns a request.Response object.
2. Add, save_args dictionary to start conforming with other datasets.
3. Raise error when trying to save non-json data.
4. Raise error for save in case of HTTP error.

## Checklist

- [x] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [x] Updated the documentation to reflect the code changes
- [x] Added a description of this change in the relevant `RELEASE.md` file
- [x] Added tests to cover my changes

### CODE REVIEW

**Comment by McDonnellJoseph** [comment]
Hello, do you know if there is documentation that needs to be updated apart from the docstring of the class ?

**Comment by McDonnellJoseph** [comment]
Thanks for pointing this out, I just added full test coverage !

**Comment by McDonnellJoseph** [comment]
Otherwise I think this is ready for review

**Comment by astrojuanlu** [comment]
@McDonnellJoseph it's an intermittent issue with unknown root cause (see gh-181), I restarted it but I'm confident the tests will pass.

**Comment by McDonnellJoseph** [comment]
@astrojuanlu  Yes that's what I thought but I don't think contributors can restart the tests. Anyway, I think the code is good for review now, got full test coverage + some better documentation for the new save feature and how it works.

File: kedro-datasets/kedro_datasets/api/api_dataset.py
Comment (noklam) [inline kedro-datasets/kedro_datasets/api/api_dataset.py:216]: So data is expected to be JSON - but the checking is missing when it is a list here?

**Review by noklam — COMMENTED** [review]
Thank you for the PR, it looks pretty solid to me!  Just thinking top of my head, does anyone ever use this for POST request? It's seem possible but I find it a bit weird. Maybe the right thing to do is `GET` for `load` and `PUT` `DELETE` `POST` for `save`?

File: kedro-datasets/kedro_datasets/api/api_dataset.py
Comment (McDonnellJoseph) [reply kedro-datasets/kedro_datasets/api/api_dataset.py:216]: At the moment, if given a list it will attempt to make a request with a slice of list of size chunk size. Do you think some kind of check should be done before making the request in the case of a list as a data input ?

Comment (McDonnellJoseph) [reply kedro-datasets/kedro_datasets/api/api_dataset.py:216]: It was also my thought at first but not requiring a specific method for load or save could allow it to work with any kind of API that suits the user ?

**Review by noklam — COMMENTED** [review]
I think it makes sense to only enable `GET` for `load` and `POST` `PUT` for `save`.  In fact we have already refactor some of the arguments in #181 (It was done in kedro develop but we forgot to sync it).

So we should have load_args and save_args, which make the dataset more consistent with other datasets too.

**Comment by noklam** [comment]
Addition - I would suggest to do the refactoring after #181 is merged as it changes the argument a bit.

**Comment by McDonnellJoseph** [comment]
> Addition - I would suggest to do the refactoring after #181 is merged as it changes the argument a bit.

Hello, maybe I misunderstood but don't you mean pr #184 ?

**Comment by noklam** [comment]
@McDonnellJoseph Sorry about that, you are right! #184 is merged now, please update this branch. There are quite are few changes because of the change in arguments. Let me know if you get stuck.

**Comment by McDonnellJoseph** [comment]
> 

Hello, I'm still waiting for a review and would be happy to make a contribution to the project. Is something missing in the PR to get a review ?

File: kedro-datasets/kedro_datasets/api/api_dataset.py
Comment (merelcht) [inline kedro-datasets/kedro_datasets/api/api_dataset.py:31]: Seems like something went wrong with formatting here. Can you revert this change?

File: kedro-datasets/kedro_datasets/api/api_dataset.py
Comment (merelcht) [inline kedro-datasets/kedro_datasets/api/api_dataset.py:55]:
```suggestion
    ``APIDataSet`` can also be used to save output on a remote server using
    HTTP(S) methods.
```

File: kedro-datasets/kedro_datasets/api/api_dataset.py
Comment (merelcht) [inline kedro-datasets/kedro_datasets/api/api_dataset.py:68]: To keep it consistent with the GB spelling elsewhere 🇬🇧 

```suggestion
    On initialisation, we can specify all the necessary parameters in the save args
```

File: kedro-datasets/kedro_datasets/api/api_dataset.py
Comment (merelcht) [inline kedro-datasets/kedro_datasets/api/api_dataset.py:72]:
```suggestion
    only used if the input of save method is a list. It will divide the request into
```

File: kedro-datasets/kedro_datasets/api/api_dataset.py
Comment (merelcht) [inline kedro-datasets/kedro_datasets/api/api_dataset.py:114]:
```suggestion
            DELETE, HEAD, etc... 
            data: The request payload, used for POST, PUT, etc
```

File: kedro-datasets/kedro_datasets/api/api_dataset.py
Comment (merelcht) [inline kedro-datasets/kedro_datasets/api/api_dataset.py:145]: It seems a bit odd that "DELETE" also means save. I think it makes more sense if we just don't support DELETE.

File: kedro-datasets/kedro_datasets/api/api_dataset.py
Comment (merelcht) [inline kedro-datasets/kedro_datasets/api/api_dataset.py:113]: Since not all methods are supported, I'd only mention the ones that are.

**Review by merelcht — COMMENTED** [review]
Thanks for the contribution @McDonnellJoseph ! I've left some comments. I'll be happy to approve when those are addressed 🙂

File: kedro-datasets/kedro_datasets/api/api_dataset.py
Comment (noklam) [reply kedro-datasets/kedro_datasets/api/api_dataset.py]: I would also be in favor of not supporting DELETE - unless someone request it since it doesn't fit into the data catalogue I/O nicely.

File: kedro-datasets/kedro_datasets/api/api_dataset.py
Comment (noklam) [inline kedro-datasets/kedro_datasets/api/api_dataset.py:30]: Maybe a typo? the format seems off and the original is correct.

File: kedro-datasets/kedro_datasets/api/api_dataset.py
Comment (noklam) [inline kedro-datasets/kedro_datasets/api/api_dataset.py:83]: I suggest starting this example with a JSON to simplify it, pandas isn't required.

File: kedro-datasets/kedro_datasets/api/api_dataset.py
Comment (noklam) [inline kedro-datasets/kedro_datasets/api/api_dataset.py:247]: Let's not support DELETE.

File: kedro-datasets/kedro_datasets/api/api_dataset.py
Comment (noklam) [inline kedro-datasets/kedro_datasets/api/api_dataset.py:210]: Can be refactor into this, `else` isn't needed.

```suggestion
        if self._request_args["method"] != "GET":
            raise DataSetError("Only GET method is supported for load")
        with sessions.Session() as session:
            return self._execute_request(session)
```

File: kedro-datasets/tests/api/test_api_dataset.py
Comment (noklam) [inline kedro-datasets/tests/api/test_api_dataset.py:217]: What is this test?

**Review by noklam — COMMENTED** [review]
@McDonnellJoseph Thanks for persevering on this. I missed the Github notification, feel free to @me or find me on Slack if I haven't responded for a long time.

**Comment by McDonnellJoseph** [comment]
Hello @noklam  @merelcht thank you for your comments I have addressed them in the latest version. I spent a lot of time this morning trying to rebase my branch to sign an old commit and this has generated a new problem meaning I can't pass the pipeline. Is this an easy fix for you ?

File: kedro-datasets/tests/api/test_api_dataset.py
Comment (astrojuanlu) [inline kedro-datasets/tests/api/test_api_dataset.py:177]: Spurious line causing a `SyntaxError`

```suggestion
    def test_proxy(self, requests_mock):
```

**Comment by astrojuanlu** [comment]
Looks like unsigned commit is 07eb097, which is already in `main`. @McDonnellJoseph let me have a quick locally at what's the right `git` spell to run to perform this rebase correctly and I'll post it here.

**Comment by McDonnellJoseph** [comment]
> Looks like unsigned commit is [07eb097](https://github.com/kedro-org/kedro-plugins/commit/07eb097c6dc8363329e46da501d4694dab35b3a6), which is already in `main`. @McDonnellJoseph let me have a quick locally at what's the right `git` spell to run to perform this rebase correctly and I'll post it here.

Thanks ! I got lost with the rebase which is the reason behind the large amount of commits for the PR

**Comment by astrojuanlu** [comment]
The conflicts here are quire hairy. Initially I managed to at least linearize the history by doing

```
git rebase -i 3b42fae --rebase-merges
```

and using this rebase-todo:

```
label onto

# Branch main
reset onto
label main

reset onto
pick 5d374cd [FEAT] add save method to APIDataset
pick 09e627e [ENH] create save_args parameter for api_dataset
pick f6147a0 [ENH] add tests for socket + http errors
pick 9580c38 [ENH] check save data is json
pick 440d249 [FIX] clean code
pick 1ae242c [ENH] handle different data types
pick ec8af04 [FIX] test coverage for exceptions
pick cb51b97 [ENH] add examples in APIDataSet docstring
fixup aa7c1fc make git happy'
pick 07eb097 sync APIDataSet  from kedro's `develop` (#184)
pick 3977607 [FIX] remove support for delete method
fixup 7374b2d [FEAT] add save method to APIDataset
drop f90c20f [FEAT] add save method to APIDataset
pick 318b90e [FIX] lint files
drop a9246d1 [FIX] fix conflicts
drop cc5a368 [FEAT] add save method to APIDataset
drop 30dfc59 [FIX] remove support for delete method
drop 9654232 [FEAT] add save method to APIDataset
pick af518cf [FIX] fix conflicts  # remove _convert_type
drop 58d217c [FIX] lint files # empty
pick d99d28c [FIX] remove fail save test
pick dab7584 [ENH] review suggestions
merge -C cf25924 main # Merge branch 'main' into dataset-api-add_save_method
pick 2ade519 [ENH] fix tests

# Rebase 3b42fae..2ade519 onto 3b42fae (41 commands)
#
# Commands:
# p, pick <commit> = use commit
# r, reword <commit> = use commit, but edit the commit message
# e, edit <commit> = use commit, but stop for amending
# s, squash <commit> = use commit, but meld into previous commit
# f, fixup [-C | -c] <commit> = like "squash" but keep only the previous
#                    commit's log message, unless -C is used, in which case
#                    keep only this commit's message; -c is same as -C but
#                    opens the editor
# x, exec <command> = run command (the rest of the line) using shell
# b, break = stop here (continue rebase later with 'git rebase --continue')
# d, drop <commit> = remove commit
# l, label <label> = label current HEAD with a name
# t, reset <label> = reset HEAD to a label
# m, merge [-C <commit> | -c <commit>] <label> [# <oneline>]
#         create a merge commit using the original merge commit's
#         message (or the oneline, if no original merge commit was
#         specified); use -c <commit> to reword the commit message
# u, update-ref <ref> = track a placeholder for the <ref> to be updated
#                       to this position in the new commits. The <ref> is
#                       updated at the end of the rebase
#
# These lines can be re-ordered; they are executed from top to bottom.
#
# If you remove a line here THAT COMMIT WILL BE LOST.
#
# However, if you remove everything, the rebase will be aborted.
#
```

After that, most of the repeated commits are dropped. However, doing a `git rebase main` or `git merge main` after that yields a lot of conflicts. I'm not sure if it's worth going through the hassle of reviewing them or if it's better to squash the result of the first interactive into one commit and just push that.

**Comment by McDonnellJoseph** [comment]
> 

Git magic 

> The conflicts here are quire hairy. Initially I managed to at least linearize the history by doing
> 
> ```
> git rebase -i 3b42fae --rebase-merges
> ```
> 
> and using this rebase-todo:
> 
> ```
> label onto
> 
> # Branch main
> reset onto
> label main
> 
> reset onto
> pick 5d374cd [FEAT] add save method to APIDataset
> pick 09e627e [ENH] create save_args parameter for api_dataset
> pick f6147a0 [ENH] add tests for socket + http errors
> pick 9580c38 [ENH] check save data is json
> pick 440d249 [FIX] clean code
> pick 1ae242c [ENH] handle different data types
> pick ec8af04 [FIX] test coverage for exceptions
> pick cb51b97 [ENH] add examples in APIDataSet docstring
> fixup aa7c1fc make git happy'
> pick 07eb097 sync APIDataSet  from kedro's `develop` (#184)
> pick 3977607 [FIX] remove support for delete method
> fixup 7374b2d [FEAT] add save method to APIDataset
> drop f90c20f [FEAT] add save method to APIDataset
> pick 318b90e [FIX] lint files
> drop a9246d1 [FIX] fix conflicts
> drop cc5a368 [FEAT] add save method to APIDataset
> drop 30dfc59 [FIX] remove support for delete method
> drop 9654232 [FEAT] add save method to APIDataset
> pick af518cf [FIX] fix conflicts  # remove _convert_type
> drop 58d217c [FIX] lint files # empty
> pick d99d28c [FIX] remove fail save test
> pick dab7584 [ENH] review suggestions
> merge -C cf25924 main # Merge branch 'main' into dataset-api-add_save_method
> pick 2ade519 [ENH] fix tests
> 
> # Rebase 3b42fae..2ade519 onto 3b42fae (41 commands)
> #
> # Commands:
> # p, pick <commit> = use commit
> # r, reword <commit> = use commit, but edit the commit message
> # e, edit <commit> = use commit, but stop for amending
> # s, squash <commit> = use commit, but meld into previous commit
> # f, fixup [-C | -c] <commit> = like "squash" but keep only the previous
> #                    commit's log message, unless -C is used, in which case
> #                    keep only this commit's message; -c is same as -C but
> #                    opens the editor
> # x, exec <command> = run command (the rest of the line) using shell
> # b, break = stop here (continue rebase later with 'git rebase --continue')
> # d, drop <commit> = remove commit
> # l, label <label> = label current HEAD with a name
> # t, reset <label> = reset HEAD to a label
> # m, merge [-C <commit> | -c <commit>] <label> [# <oneline>]
> #         create a merge commit using the original merge commit's
> #         message (or the oneline, if no original merge commit was
> #         specified); use -c <commit> to reword the commit message
> # u, update-ref <ref> = track a placeholder for the <ref> to be updated
> #                       to this position in the new commits. The <ref> is
> #                       updated at the end of the rebase
> #
> # These lines can be re-ordered; they are executed from top to bottom.
> #
> # If you remove a line here THAT COMMIT WILL BE LOST.
> #
> # However, if you remove everything, the rebase will be aborted.
> #
> ```
> 
> After that, most of the repeated commits are dropped. However, doing a `git rebase main` or `git merge main` after that yields a lot of conflicts. I'm not sure if it's worth going through the hassle of reviewing them or if it's better to squash the result of the first interactive into one commit and just push that.

Git magic worked ! Thanks a lot, won't the commits be squashed once the branch is merged ?

**Comment by astrojuanlu** [comment]
👽 git magic is indeed miraculous!

Yes, the commits will be squashed on merge. But to achieve that, even with this nice linearized history, first you'll need to fix the current conflicts.

**Comment by astrojuanlu** [comment]
Well done @McDonnellJoseph, thanks for persevering! 🙌🏽 I think this is ready for another round of reviews.

File: kedro-datasets/kedro_datasets/api/api_dataset.py
Comment (noklam) [inline kedro-datasets/kedro_datasets/api/api_dataset.py:30]: I still think this is wrong? Did I misunderstand something

File: kedro-datasets/kedro_datasets/api/api_dataset.py
Comment (noklam) [inline kedro-datasets/kedro_datasets/api/api_dataset.py:76]:
```suggestion
    defines how long  our program waits for a response after a request. ``chunk_size``, is
```

File: kedro-datasets/kedro_datasets/api/api_dataset.py
Comment (noklam) [inline kedro-datasets/kedro_datasets/api/api_dataset.py:75]:
```suggestion
    Two important parameters to keep in mind are `timeout` and `chunk_size`.    
```

Does this render into something special? There are many capitalised version of the parameters. If it's a variable then use `` `` to quote it.

File: kedro-datasets/kedro_datasets/api/api_dataset.py
Comment (noklam) [inline kedro-datasets/kedro_datasets/api/api_dataset.py:82]: It would be great to make sure the docstring does work when you copy and paste, I believe it's not a table anymore and `to_dict` is not needed.

File: kedro-datasets/kedro_datasets/api/api_dataset.py
Comment (noklam) [inline kedro-datasets/kedro_datasets/api/api_dataset.py:112]:
```suggestion
            method: The method of the request. GET, POST, PUT are the only supported
```

Comment (noklam) [inline kedro-datasets/kedro_datasets/api/api_dataset.py:112]:
```suggestion
            method: The method of the request. GET, POST, PUT are the only supported
```

**Review by noklam — COMMENTED** [review]
Thank you, the PR is looking better now.

If the DCO fails, I suggested just leave it until everything is reviewed, as this is a little bit hard to review now since all previous review comments are gone.

In general it's good to have documentation consistent with other dataset, I would suggest to have a look at the popular one (i.e. CSVDataSet).

**Review by noklam — COMMENTED** [review]
Thank you, the PR is looking better now.

If the DCO fails, I suggested just leave it until everything is reviewed, as this is a little bit hard to review now since all previous review comments are gone.

In general it's good to have documentation consistent with other dataset, I would suggest to have a look at the popular one (i.e. CSVDataSet).

**Comment by McDonnellJoseph** [comment]
> Thank you, the PR is looking better now.
> 
> If the DCO fails, I suggested just leave it until everything is reviewed, as this is a little bit hard to review now since all previous review comments are gone.
> 
> In general it's good to have documentation consistent with other dataset, I would suggest to have a look at the popular one (i.e. CSVDataSet).

Do you think there is somewhere more appropriate in the file for the explanation paragraph on how the save method works ? Or should I just dump it altogether ?

File: kedro-datasets/kedro_datasets/api/api_dataset.py
Comment (McDonnellJoseph) [reply kedro-datasets/kedro_datasets/api/api_dataset.py:30]: my editor keeps messing this line

**Comment by noklam** [comment]
I think it's fine to leave it within the docstring, @stichbury for comments

**Review by astrojuanlu — APPROVED** [review]
Well done @McDonnellJoseph!

**Review by noklam — APPROVED** [review]

**Comment by astrojuanlu** [comment]
Congrats @McDonnellJoseph and thank you for your contribution! 🙌🏽

**Pull request merged**
